### ISSUE TITLE
[BUG] High latency, even with USB tethering

### ISSUE THREAD

**Issue by Kamilkampfwagen-II** [mkckr0/audio-share#16]
**Describe the bug**
I've connected the phone to PC over USB 2.0 and enabled Tethering over USB to have lower latency.
I've picked the correct adapter's IP (not the ethernet cable or anything). I also disabled any matters of wireless connectivity on the phone.
Pinging each other from each end, the average latency is around `0.5ms` (half a milisecond).
However there's a significant delay in audio of around 350ms, although my phone is quite old (2015, low-end), I am still not sure if it is purely the decompression overhead on its end.

**To Reproduce**
1. Connect phone to PC over USB
2. Enable tethering over USB on Android
3. Start the server with the default endpoint which regular speakers are connected to. Make sure to pick the correct IP address corresponding to the adapter newly created.
4. Using this [YouTube video](https://youtu.be/ucZl6vQ_8Uo), determine the average delay in audio

**Expected behavior**
Way lower amount of latency

**Server:**
 - OS: Windows 10 21H2 19044.3570
 - CMD
 - Version: v0.0.11

**Android:**
 - Device: ASUS Zenfone 2 Laser
 - OS: Android 7 (LineageOS custom ROM)
 - Version: v0.0.11

**Additional information**
```
> audio-share-server.exe -b 192.168.42.182
[2023-11-05 14:39:38.685] [info] endpoint_id: {0.0.0.00000000}.{9b8dcbfc-a8ce-41ad-87c6-efa0e11b319a}
[2023-11-05 14:39:38.685] [info] tcp listen success on 192.168.42.182:65530
[2023-11-05 14:39:38.686] [info] udp listen success on 192.168.42.182:65530
[2023-11-05 14:39:38.686] [info] server started
[2023-11-05 14:39:38.686] [info] select audio endpoint: Hoparl÷r (Realtek(R) Audio)
[2023-11-05 14:39:38.690] [info] WAVEFORMATEX: wFormatTag: 65534, nBlockAlign: 8
[2023-11-05 14:39:38.690] [info] AudioFormat:
format_tag: 3
channels: 2
sample_rate: 48000
bits_per_sample: 32

[2023-11-05 14:39:55.602] [info] accept 192.168.42.129:49387

```

**Comment by mkckr0** [comment]
> Pinging each other from each end, the average latency is around 0.5ms (half a milisecond).

How do you ping phone? I test the ping delay is about 2~5ms.

Are both delay of via WIFI or USB tethering all reach 350ms?

**Comment by Kamilkampfwagen-II** [comment]
> How do you ping phone? I test the ping delay is about 2~5ms.

I do get the local IP address of the phone from the server logs when the phone is connected. Additionally it can also be gathered using termux with the `ifconfig` command.

On PC I use the ping command. It doesn't show sub 1ms values in Windows but it does on Linux (so on the phone)


> Do delay of via WIFI and USB tethering both reach 350ms?

Here's a table:
| Latency: | USB      | Wi-Fi    |
|----------|----------|----------|
| Ping     | `~0.5ms` | `2~5ms`  |
| Audio    | `~350ms` | `~500ms` |

These values are pretty consistent even over long periods
The measurements in audio delay is pretty loose since I don't have a reliable method other than using the video I referenced. However even assuming a worst case `100ms` error margin, the latency over USB is quite high.

**Comment by mkckr0** [comment]
There is no decompression or any other complicated jobs. It's just memory copy. The delay may be caused by the CPU performance and RAM R/W delay. You can use another PC or phone to test. I don't have better idea to figure it out.

**Comment by mkckr0** [comment]
Your problem may be same as #14. You can use `wav_recorder.exe` to test if exists delay.
[wav_recorder.zip](https://github.com/mkckr0/audio-share/files/13213022/wav_recorder.zip)
https://github.com/mkckr0/audio-share/issues/14#issuecomment-1786558474

**Comment by mkckr0** [comment]
Fix at #24.
